File: packages/firebase_analytics/firebase_analytics/android/src/main/java/io/flutter/plugins/firebase/analytics/FlutterFirebaseAnalyticsPlugin.java
// Copyright 2017 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package io.flutter.plugins.firebase.analytics;

import android.content.Context;
import android.os.Bundle;
import android.os.Parcelable;
import androidx.annotation.NonNull;
import com.google.android.gms.tasks.Task;
import com.google.android.gms.tasks.TaskCompletionSource;
import com.google.android.gms.tasks.Tasks;
import com.google.firebase.FirebaseApp;
import com.google.firebase.analytics.FirebaseAnalytics;
import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugins.firebase.core.FlutterFirebasePlugin;
import io.flutter.plugins.firebase.core.FlutterFirebasePluginRegistry;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

/** Flutter plugin for Firebase Analytics. */
public class FlutterFirebaseAnalyticsPlugin
    implements FlutterFirebasePlugin, MethodCallHandler, FlutterPlugin {
  private FirebaseAnalytics analytics;
  private MethodChannel channel;

  private void initInstance(BinaryMessenger messenger, Context context) {
    analytics = FirebaseAnalytics.getInstance(context);
    String channelName = "plugins.flutter.io/firebase_analytics";
    channel = new MethodChannel(messenger, channelName);
    channel.setMethodCallHandler(this);
    FlutterFirebasePluginRegistry.registerPlugin(channelName, this);
  }

  @SuppressWarnings("unchecked")
  private static Bundle createBundleFromMap(Map<String, Object> map) {
    if (map == null) {
      return null;
    }

    Bundle bundle = new Bundle();
    for (Map.Entry<String, Object> jsonParam : map.entrySet()) {
      final Object value = jsonParam.getValue();
      final String key = jsonParam.getKey();
      if (value instanceof String) {
        bundle.putString(key, (String) value);
      } else if (value instanceof Integer) {
        // FirebaseAnalytics default event parameters only support long and double types, so we convert the int to a long.
        bundle.putLong(key, (Integer) value);
      } else if (value instanceof Long) {
        bundle.putLong(key, (Long) value);
      } else if (value instanceof Double) {
        bundle.putDouble(key, (Double) value);
      } else if (value instanceof Boolean) {
        bundle.putBoolean(key, (Boolean) value);
      } else if (value == null) {
        bundle.putString(key, null);
      } else if (value instanceof Iterable<?>) {
        ArrayList<Parcelable> list = new ArrayList<>();

        for (Object item : (Iterable<?>) value) {
          if (item instanceof Map) {
            //noinspection unchecked
            list.add(createBundleFromMap((Map<String, Object>) item));
          } else {
            throw new IllegalArgumentException(
                "Unsupported value type: "
                    + item.getClass().getCanonicalName()
                    + " in list at key "
                    + key);
          }
        }

        bundle.putParcelableArrayList(key, list);
      } else if (value instanceof Map<?, ?>) {
        //noinspection unchecked
        bundle.putParcelable(key, createBundleFromMap((Map<String, Object>) value));
      } else {
        throw new IllegalArgumentException(
            "Unsupported value type: " + value.getClass().getCanonicalName());
      }
    }
    return bundle;
  }

  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding binding) {
    initInstance(binding.getBinaryMessenger(), binding.getApplicationContext());
  }

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    if (channel != null) {
      channel.setMethodCallHandler(null);
      channel = null;
    }
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
    Task<?> methodCallTask;

    switch (call.method) {
      case "Analytics#logEvent":
        methodCallTask = handleLogEvent(call.arguments());
        break;
      case "Analytics#setUserId":
        methodCallTask = handleSetUserId(call.arguments());
        break;
      case "Analytics#setAnalyticsCollectionEnabled":
        methodCallTask = handleSetAnalyticsCollectionEnabled(call.arguments());
        break;
      case "Analytics#setSessionTimeoutDuration":
        methodCallTask = handleSetSessionTimeoutDuration(call.arguments());
        break;
      case "Analytics#setUserProperty":
        methodCallTask = handleSetUserProperty(call.arguments());
        break;
      case "Analytics#resetAnalyticsData":
        methodCallTask = handleResetAnalyticsData();
        break;
      case "Analytics#setConsent":
        methodCallTask = setConsent(call.arguments());
        break;
      case "Analytics#setDefaultEventParameters":
        methodCallTask = setDefaultEventParameters(call.arguments());
        break;
      case "Analytics#getAppInstanceId":
        methodCallTask = handleGetAppInstanceId();
        break;
      default:
        result.notImplemented();
        return;
    }

    methodCallTask.addOnCompleteListener(
        task -> {
          if (task.isSuccessful()) {
            result.success(task.getResult());
          } else {
            Exception exception = task.getException();
            String message =
                exception != null ? exception.getMessage() : "An unknown error occurred";
            result.error("firebase_analytics", message, null);
          }
        });
  }

  private Task<Void> handleLogEvent(final Map<String, Object> arguments) {
    TaskCompletionSource<Void> taskCompletionSource = new TaskCompletionSource<>();

    cachedThreadPool.execute(
        () -> {
          try {
            final String eventName =
                (String) Objects.requireNonNull(arguments.get(Constants.EVENT_NAME));
            @SuppressWarnings("unchecked")
            final Map<String, Object> map =
                (Map<String, Object>) arguments.get(Constants.PARAMETERS);
            final Bundle parameterBundle = createBundleFromMap(map);
            analytics.logEvent(eventName, parameterBundle);
            taskCompletionSource.setResult(null);
          } catch (Exception e) {
            taskCompletionSource.setException(e);
          }
        });

    return taskCompletionSource.getTask();
  }

  private Task<Void> handleSetUserId(final Map<String, Object> arguments) {
    TaskCompletionSource<Void> taskCompletionSource = new TaskCompletionSource<>();

    cachedThreadPool.execute(
        () -> {
          try {
            final String id = (String) arguments.get(Constants.USER_ID);
            analytics.setUserId(id);
            taskCompletionSource.setResult(null);
          } catch (Exception e) {
            taskCompletionSource.setException(e);
          }
        });

    return taskCompletionSource.getTask();
  }

  private Task<Void> handleSetAnalyticsCollectionEnabled(final Map<String, Object> arguments) {
    TaskCompletionSource<Void> taskCompletionSource = new TaskCompletionSource<>();

    cachedThreadPool.execute(
        () -> {
          try {
            final Boolean enabled =
                (Boolean) Objects.requireNonNull(arguments.get(Constants.ENABLED));
            analytics.setAnalyticsCollectionEnabled(enabled);
            taskCompletionSource.setResult(null);
          } catch (Exception e) {
            taskCompletionSource.setException(e);
          }
        });

    return taskCompletionSource.getTask();
  }

  private Task<Void> handleSetSessionTimeoutDuration(final Map<String, Object> arguments) {
    TaskCompletionSource<Void> taskCompletionSource = new TaskCompletionSource<>();

    cachedThreadPool.execute(
        () -> {
          try {
            final Integer milliseconds =
                (Integer) Objects.requireNonNull(arguments.get(Constants.MILLISECONDS));
            analytics.setSessionTimeoutDuration(milliseconds);
            taskCompletionSource.setResult(null);
          } catch (Exception e) {
            taskCompletionSource.setException(e);
          }
        });

    return taskCompletionSource.getTask();
  }

  private Task<Void> handleSetUserProperty(final Map<String, Object> arguments) {
    TaskCompletionSource<Void> taskCompletionSource = new TaskCompletionSource<>();

    cachedThreadPool.execute(
        () -> {
          try {
            final String name = (String) Objects.requireNonNull(arguments.get(Constants.NAME));
            final String value = (String) arguments.get(Constants.VALUE);
            analytics.setUserProperty(name, value);
            taskCompletionSource.setResult(null);
          } catch (Exception e) {
            taskCompletionSource.setException(e);
          }
        });

    return taskCompletionSource.getTask();
  }

  private Task<Void> handleResetAnalyticsData() {
    TaskCompletionSource<Void> taskCompletionSource = new TaskCompletionSource<>();

    cachedThreadPool.execute(
        () -> {
          try {
            analytics.resetAnalyticsData();
            taskCompletionSource.setResult(null);
          } catch (Exception e) {
            taskCompletionSource.setException(e);
          }
        });

    return taskCompletionSource.getTask();
  }

  private Task<Void> setConsent(final Map<String, Object> arguments) {
    TaskCompletionSource<Void> taskCompletionSource = new TaskCompletionSource<>();

    cachedThreadPool.execute(
        () -> {
          try {
            final Boolean adStorageGranted =
                (Boolean) arguments.get(Constants.AD_STORAGE_CONSENT_GRANTED);
            final Boolean analyticsStorageGranted =
                (Boolean) arguments.get(Constants.ANALYTICS_STORAGE_CONSENT_GRANTED);
            HashMap<FirebaseAnalytics.ConsentType, FirebaseAnalytics.ConsentStatus> parameters =
                new HashMap<>();

            if (adStorageGranted != null) {
              parameters.put(
                  FirebaseAnalytics.ConsentType.AD_STORAGE,
                  adStorageGranted
                      ? FirebaseAnalytics.ConsentStatus.GRANTED
                      : FirebaseAnalytics.ConsentStatus.DENIED);
            }

            if (analyticsStorageGranted != null) {
              parameters.put(
                  FirebaseAnalytics.ConsentType.ANALYTICS_STORAGE,
                  analyticsStorageGranted
                      ? FirebaseAnalytics.ConsentStatus.GRANTED
                      : FirebaseAnalytics.ConsentStatus.DENIED);
            }

            analytics.setConsent(parameters);
            taskCompletionSource.setResult(null);
          } catch (Exception e) {
            taskCompletionSource.setException(e);
          }
        });

    return taskCompletionSource.getTask();
  }

  private Task<Void> setDefaultEventParameters(final Map<String, Object> arguments) {
    TaskCompletionSource<Void> taskCompletionSource = new TaskCompletionSource<>();

    cachedThreadPool.execute(
        () -> {
          try {
            analytics.setDefaultEventParameters(createBundleFromMap(arguments));
            taskCompletionSource.setResult(null);
          } catch (Exception e) {
            taskCompletionSource.setException(e);
          }
        });

    return taskCompletionSource.getTask();
  }

  private Task<String> handleGetAppInstanceId() {
    TaskCompletionSource<String> taskCompletionSource = new TaskCompletionSource<>();

    cachedThreadPool.execute(
        () -> {
          try {
            taskCompletionSource.setResult(Tasks.await(analytics.getAppInstanceId()));
          } catch (Exception e) {
            taskCompletionSource.setException(e);
          }
        });

    return taskCompletionSource.getTask();
  }

  @Override
  public Task<Map<String, Object>> getPluginConstantsForFirebaseApp(FirebaseApp firebaseApp) {
    TaskCompletionSource<Map<String, Object>> taskCompletionSource = new TaskCompletionSource<>();

    cachedThreadPool.execute(
        () -> {
          try {
            taskCompletionSource.setResult(new HashMap<String, Object>() {});
          } catch (Exception e) {
            taskCompletionSource.setException(e);
          }
        });

    return taskCompletionSource.getTask();
  }

  @Override
  public Task<Void> didReinitializeFirebaseCore() {
    TaskCompletionSource<Void> taskCompletionSource = new TaskCompletionSource<>();

    cachedThreadPool.execute(
        () -> {
          try {
            taskCompletionSource.setResult(null);
          } catch (Exception e) {
            taskCompletionSource.setException(e);
          }
        });

    return taskCompletionSource.getTask();
  }
}

File: packages/firebase_analytics/firebase_analytics/android/src/main/kotlin/io/flutter/plugins/firebase/analytics/FlutterFirebaseAnalyticsPlugin.kt
// Copyright 2018 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package io.flutter.plugins.firebase.analytics

import android.content.Context
import android.os.Bundle
import android.os.Parcelable
import com.google.android.gms.tasks.Task
import com.google.android.gms.tasks.TaskCompletionSource
import com.google.android.gms.tasks.Tasks
import com.google.firebase.FirebaseApp
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.firebase.analytics.FirebaseAnalytics.ConsentStatus
import com.google.firebase.analytics.FirebaseAnalytics.ConsentType
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugins.firebase.core.FlutterFirebasePlugin
import io.flutter.plugins.firebase.core.FlutterFirebasePlugin.cachedThreadPool
import io.flutter.plugins.firebase.core.FlutterFirebasePluginRegistry
import java.util.Objects

class FlutterFirebaseAnalyticsPlugin : FlutterFirebasePlugin,
  FlutterPlugin, FirebaseAnalyticsHostApi {
  private lateinit var analytics: FirebaseAnalytics
  private var channel: MethodChannel? = null

  private var messenger: BinaryMessenger? = null


  private fun initInstance(messenger: BinaryMessenger, context: Context) {
    analytics = FirebaseAnalytics.getInstance(context)
    val channelName = "plugins.flutter.io/firebase_analytics"
    channel = MethodChannel(messenger, channelName)
    FirebaseAnalyticsHostApi.setUp(messenger, this)
    FlutterFirebasePluginRegistry.registerPlugin(channelName, this)
    this.messenger = messenger
  }

  override fun getPluginConstantsForFirebaseApp(firebaseApp: FirebaseApp?): Task<MutableMap<String, Any>> {
    val taskCompletionSource = TaskCompletionSource<MutableMap<String, Any>>()

    cachedThreadPool.execute {
      try {
        taskCompletionSource.setResult(HashMap())
      } catch (e: Exception) {
        taskCompletionSource.setException(e)
      }
    }

    return taskCompletionSource.task
  }

  override fun didReinitializeFirebaseCore(): Task<Void> {
    val taskCompletionSource = TaskCompletionSource<Void>()

    cachedThreadPool.execute {
      try {
        taskCompletionSource.setResult(null)
      } catch (e: java.lang.Exception) {
        taskCompletionSource.setException(e)
      }
    }

    return taskCompletionSource.task
  }

  override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    initInstance(binding.binaryMessenger, binding.applicationContext)
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel?.setMethodCallHandler(null)

    checkNotNull(messenger)
    FirebaseAnalyticsHostApi.setUp(messenger!!, null)

    channel = null
    messenger = null
  }

  private fun handleGetSessionId(): Task<Long> {
    val taskCompletionSource = TaskCompletionSource<Long>()

    cachedThreadPool.execute {
      try {
        taskCompletionSource.setResult(
          Tasks.await(
            analytics.sessionId
          )
        )
      } catch (e: java.lang.Exception) {
        taskCompletionSource.setException(e)
      }
    }

    return taskCompletionSource.task
  }

  private fun handleLogEvent(arguments: Map<String, Any?>): Task<Void> {
    val taskCompletionSource = TaskCompletionSource<Void>()

    cachedThreadPool.execute {
      try {
        val eventName =
          Objects.requireNonNull(arguments[Constants.EVENT_NAME]) as String
        val map =
          arguments[Constants.PARAMETERS] as Map<String, Any>?
        val parameterBundle: Bundle? =
          createBundleFromMap(
            map
          )
        analytics.logEvent(eventName, parameterBundle)
        taskCompletionSource.setResult(null)
      } catch (e: java.lang.Exception) {
        taskCompletionSource.setException(e)
      }
    }

    return taskCompletionSource.task
  }

  private fun handleSetUserId(userId: String?): Task<Void> {
    val taskCompletionSource = TaskCompletionSource<Void>()

    cachedThreadPool.execute {
      try {
        analytics.setUserId(userId)
        taskCompletionSource.setResult(null)
      } catch (e: java.lang.Exception) {
        taskCompletionSource.setException(e)
      }
    }

    return taskCompletionSource.task
  }

  private fun handleSetAnalyticsCollectionEnabled(enabled: Boolean): Task<Void> {
    val taskCompletionSource = TaskCompletionSource<Void>()

    cachedThreadPool.execute {
      try {
        analytics.setAnalyticsCollectionEnabled(enabled)
        taskCompletionSource.setResult(null)
      } catch (e: java.lang.Exception) {
        taskCompletionSource.setException(e)
      }
    }

    return taskCompletionSource.task
  }

  private fun handleSetSessionTimeoutDuration(milliseconds: Long): Task<Void> {
    val taskCompletionSource = TaskCompletionSource<Void>()

    cachedThreadPool.execute {
      try {
        analytics.setSessionTimeoutDuration(milliseconds)
        taskCompletionSource.setResult(null)
      } catch (e: java.lang.Exception) {
        taskCompletionSource.setException(e)
      }
    }

    return taskCompletionSource.task
  }

  private fun handleSetUserProperty(name: String, value: String?): Task<Void> {
    val taskCompletionSource = TaskCompletionSource<Void>()

    cachedThreadPool.execute {
      try {
        analytics.setUserProperty(name, value)
        taskCompletionSource.setResult(null)
      } catch (e: java.lang.Exception) {
        taskCompletionSource.setException(e)
      }
    }

    return taskCompletionSource.task
  }

  private fun handleResetAnalyticsData(): Task<Void> {
    val taskCompletionSource = TaskCompletionSource<Void>()

    cachedThreadPool.execute {
      try {
        analytics.resetAnalyticsData()
        taskCompletionSource.setResult(null)
      } catch (e: java.lang.Exception) {
        taskCompletionSource.setException(e)
      }
    }

    return taskCompletionSource.task
  }

  private fun handleSetConsent(arguments: Map<String, Boolean?>): Task<Void> {
    val taskCompletionSource = TaskCompletionSource<Void>()

    cachedThreadPool.execute {
      try {
        val adStorageGranted =
          arguments[Constants.AD_STORAGE_CONSENT_GRANTED]
        val analyticsStorageGranted =
          arguments[Constants.ANALYTICS_STORAGE_CONSENT_GRANTED]
        val adPersonalizationSignalsGranted =
          arguments[Constants.AD_PERSONALIZATION_SIGNALS_CONSENT_GRANTED]
        val adUserDataGranted =
          arguments[Constants.AD_USER_DATA_CONSENT_GRANTED]
        val parameters =
          java.util.HashMap<ConsentType, ConsentStatus>()

        if (adStorageGranted != null) {
          parameters[ConsentType.AD_STORAGE] = if (adStorageGranted)
            ConsentStatus.GRANTED
          else
            ConsentStatus.DENIED
        }

        if (analyticsStorageGranted != null) {
          parameters[ConsentType.ANALYTICS_STORAGE] = if (analyticsStorageGranted)
            ConsentStatus.GRANTED
          else
            ConsentStatus.DENIED
        }

        if (adPersonalizationSignalsGranted != null) {
          parameters[ConsentType.AD_PERSONALIZATION] =
            if (adPersonalizationSignalsGranted)
              ConsentStatus.GRANTED
            else
              ConsentStatus.DENIED
        }

        if (adUserDataGranted != null) {
          parameters[ConsentType.AD_USER_DATA] = if (adUserDataGranted)
            ConsentStatus.GRANTED
          else
            ConsentStatus.DENIED
        }

        analytics.setConsent(parameters)
        taskCompletionSource.setResult(null)
      } catch (e: java.lang.Exception) {
        taskCompletionSource.setException(e)
      }
    }

    return taskCompletionSource.task
  }

  private fun handleSetDefaultEventParameters(parameters: Map<String, Any?>?): Task<Void> {
    val taskCompletionSource = TaskCompletionSource<Void>()

    cachedThreadPool.execute {
      try {
        analytics.setDefaultEventParameters(
          createBundleFromMap(
            parameters
          )
        )
        taskCompletionSource.setResult(null)
      } catch (e: java.lang.Exception) {
        taskCompletionSource.setException(e)
      }
    }

    return taskCompletionSource.task
  }

  private fun handleGetAppInstanceId(): Task<String> {
    val taskCompletionSource = TaskCompletionSource<String>()

    cachedThreadPool.execute {
      try {
        taskCompletionSource.setResult(
          Tasks.await(
            analytics.appInstanceId
          )
        )
      } catch (e: java.lang.Exception) {
        taskCompletionSource.setException(e)
      }
    }

    return taskCompletionSource.task
  }

  private fun createBundleFromMap(map: Map<String, Any?>?): Bundle? {
    if (map == null) {
      return null
    }

    val bundle = Bundle()
    for ((key, value) in map) {
      if (value is String) {
        bundle.putString(key, value)
      } else if (value is Int) {
        // FirebaseAnalytics default event parameters only support long and double types, so we convert the int to a long.
        bundle.putLong(key, value.toLong())
      } else if (value is Long) {
        bundle.putLong(key, value)
      } else if (value is Double) {
        bundle.putDouble(key, value)
      } else if (value is Boolean) {
        bundle.putBoolean(key, value)
      } else if (value == null) {
        bundle.putString(key, null)
      } else if (value is Iterable<*>) {
        val list = ArrayList<Parcelable?>()

        for (item in value) {
          if (item is Map<*, *>) {
            list.add(createBundleFromMap(item as Map<String, Any>))
          } else {
            if (item != null) {
              throw IllegalArgumentException(
                ("Unsupported value type: "
                  + item.javaClass.canonicalName
                  + " in list at key "
                  + key)
              )
            }
          }
        }

        bundle.putParcelableArrayList(key, list)
      } else if (value is Map<*, *>) {
        bundle.putParcelable(key, createBundleFromMap(value as Map<String, Any>))
      } else {
        throw IllegalArgumentException(
          "Unsupported value type: " + value.javaClass.canonicalName
        )
      }
    }
    return bundle
  }

  private fun handleVoidTaskResult(
    task: Task<Void?>,
    callback: (Result<Unit>) -> Unit
  ) {
    if (task.isSuccessful) {
      callback(Result.success(Unit))
    } else {
      val message = task.exception?.message ?: "An unknown error occurred"
      callback(Result.failure(FlutterError("firebase_analytics", message, null)))
    }
  }

  private fun <T> handleTypedTaskResult(
    task: Task<T>,
    callback: (Result<T?>) -> Unit
  ) {
    if (task.isSuccessful) {
      callback(Result.success(task.result))
    } else {
      val message = task.exception?.message ?: "An unknown error occurred"
      callback(Result.failure(FlutterError("firebase_analytics", message, null)))
    }
  }

  override fun logEvent(event: Map<String, Any?>, callback: (Result<Unit>) -> Unit) {
    handleLogEvent(event).addOnCompleteListener { task ->
      handleVoidTaskResult(task, callback)
    }
  }


  override fun setUserId(userId: String?, callback: (Result<Unit>) -> Unit) {
    handleSetUserId(userId).addOnCompleteListener { task ->
      handleVoidTaskResult(task, callback)
    }
  }

  override fun setUserProperty(name: String, value: String?, callback: (Result<Unit>) -> Unit) {
    handleSetUserProperty(name, value).addOnCompleteListener { task ->
      handleVoidTaskResult(task, callback)
    }
  }

  override fun setAnalyticsCollectionEnabled(enabled: Boolean, callback: (Result<Unit>) -> Unit) {
    handleSetAnalyticsCollectionEnabled(enabled).addOnCompleteListener { task ->
      handleVoidTaskResult(task, callback)
    }
  }

  override fun resetAnalyticsData(callback: (Result<Unit>) -> Unit) {
    handleResetAnalyticsData().addOnCompleteListener { task ->
      handleVoidTaskResult(task, callback)
    }
  }

  override fun setSessionTimeoutDuration(timeout: Long, callback: (Result<Unit>) -> Unit) {
    handleSetSessionTimeoutDuration(timeout).addOnCompleteListener { task ->
      handleVoidTaskResult(task, callback)
    }
  }

  override fun setConsent(consent: Map<String, Boolean?>, callback: (Result<Unit>) -> Unit) {
    handleSetConsent(consent).addOnCompleteListener { task ->
      handleVoidTaskResult(task, callback)
    }

  }

  override fun setDefaultEventParameters(
    parameters: Map<String, Any?>?,
    callback: (Result<Unit>) -> Unit
  ) {
    handleSetDefaultEventParameters(parameters).addOnCompleteListener { task ->
      handleVoidTaskResult(task, callback)
    }
  }


  override fun getAppInstanceId(callback: (Result<String?>) -> Unit) {
    handleGetAppInstanceId().addOnCompleteListener { task ->
      handleTypedTaskResult(task, callback)
    }
  }

  override fun getSessionId(callback: (Result<Long?>) -> Unit) {
    handleGetSessionId().addOnCompleteListener { task ->
      handleTypedTaskResult(task, callback)
    }
  }

  override fun initiateOnDeviceConversionMeasurement(
    arguments: Map<String, String?>,
    callback: (Result<Unit>) -> Unit
  ) {
    callback(
      Result.failure(
        FlutterError(
          "unimplemented",
          "initiateOnDeviceConversionMeasurement is only available on iOS.",
          null
        )
      )
    )
  }
}
